substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/image/NativeImageDebugInfoProvider.java

```diff
@@ -309,8 +314,9 @@ private void computeFullFilePathAndCachePath() {
                 clazz = ((OriginalClassProvider) declaringClass).getJavaClass();
             }
             /*
-             * HostedType and AnalysisType punt calls to getSourceFilename to the wrapped class so
-             * for consistency we need to do the path lookup relative to the wrapped class.
+             * HostedType wraps an AnalysisType and both HostedType and AnalysisType punt calls to
+             * getSourceFilename to the wrapped class so for consistency we need to do the path
+             * lookup relative to the doubly unwrapped HostedType.
              */
             if (declaringClass instanceof HostedType) {
                 declaringClass = ((HostedType) declaringClass).getWrapped();
```

Comment:
Shouldn't this be:

```suggestion
                declaringClass = ((HostedType) declaringClass).getWrapped().getWrapped();
```
?

Reply:
No, I don't think so. The result of that single getWrapped() call is an AnalysisType. The next if detects that and unwraps a second time. That second if also deals with the case where an AnalysisType comes in direct.

Reply:
Indeed. So a potential optimization would be to change the following `if` in an `else if` since the result of `HostedType#getWrapped()` returns an `AnalysisType` as you point out.

I might be nit-picking here :)

Reply:
On a second thought, I think this should be fixed, because as it stands it only works because the if statements are in the correct order, making it easy to break things with a simple reordering without any clear indication that these two  statements should not be reordered.

Reply:
The if cascade is correct and the comment provides the information which explains why. However, I agree it does not spell it out. I will change the comment to make this easier to infer:

            /*
             * HostedType wraps an AnalysisType and both HostedType and AnalysisType punt calls to
             * getSourceFilename to the wrapped class so for consistency we need to do the path
             * lookup relative to the doubly unwrapped HostedType or singly unwrapped AnalysisType.
             */
I'm not sure we need to protect any further against programmers who transpose conditional operations that side-effect the elements of their conditions. Instead I recommend we just reject their contributions. ;-)